clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1072,12 +1073,10 @@ public Map<TopicPartition, OffsetAndTimestamp> offsetsForTimes(Map<TopicPartitio
             }
 
             try {
-                return applicationEventHandler.addAndGet(listOffsetsEvent)
-                    .entrySet()
-                    .stream()
-                    .collect(Collectors.toMap(
-                        Map.Entry::getKey,
-                        entry -> entry.getValue().buildOffsetAndTimestamp()));
+                Map<TopicPartition, OffsetAndTimestampInternal> offsets = applicationEventHandler.addAndGet(listOffsetsEvent);
+                Map<TopicPartition, OffsetAndTimestamp> results = new HashMap<>(offsets.size());
+                offsets.forEach((k, v) -> results.put(k, v != null ? v.buildOffsetAndTimestamp() : null));
```

Comment:
The documentation does state that `{@code null} will be returned for the partition if there is no such message`, but it feels odd to insert a null as the value.

If it's necessary to align the behavior with `ClassicKafkaConsumer`, the other methods like `beginningOffsets` and `endOffsets` might also encounter a similar issue with the `null` value.

https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1132

Reply:
>  What if the user calls Map.containsKey(), though? Without explicitly adding a null, containsKey() will return false. Is that the expectation? 🤔

I'm not sure why users would need this use case. Storing null in a map seems like a design flaw to me. Additionally, Admin#listOffsets doesn't return null values.

Reply:
[`OffsetFetchUtils.buildListOffsetsResult()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java#L243-L258) is used by the `ClassicKafkaConsumer` when building up the results for `offsetsForTimes()`. That method has two loops—the first loops over the `TopicPartition`s that were _requested_ and fills in `offsetsResults` with `null`. The second loops over the `TopicPartition`s with _results_ and replaces the `null` with the `OffsetAndTimestamp`. The idea was to mimic that in the `AsyncKafkaConsumer` implementation.

Thoughts?

Reply:
> If it's necessary to align the behavior with `ClassicKafkaConsumer`, the other methods like `beginningOffsets` and `endOffsets` might also encounter a similar issue with the `null` value.

Yes, I noticed that, but I thought those didn't need to change...

Interestingly, [`OffsetFetcher.beginningOrEndOffset()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcher.java#L194-L210) (used by `ClassicKafkaConsumer`) does _not_ pre-populate the returned hash table with `null`s. As a result, `beginningOffsets()` and `endOffsets()` for the `ClassicKafkaConsumer` doesn't have to worry about potential `NullPointerExceptions`. So I'd decided that the `AsyncKafkaConsumer`'s corresponding methods also didn't have the same potential for `NullPointerException`s.

_However_, I just double-checked, and it turns out that `AsyncKafkaConsumer` is using a lot of the same underlying code for `beginningOffsets()` and `endOffsets()` as it does for `offsetsForTimes()`. So the new consumer is also pre-populating the returned hash tables for `beginningOffsets()` and `endOffsets()` with `null`, which is not consistent with the old consumer.

I'll add some unit tests to trigger these cases and make sure `AsyncKafkaConsumer` is consistent with `ClassicKafkaConsumer`, warts and all.

Reply:
> The idea was to mimic that in the AsyncKafkaConsumer implementation.

I understand the intention to mimic the existing behavior, but do we really need to mimic a bad design? Especially since newer Java collection operations, like lambdas and `Map.copyOf`, don't accept `null` values. In my opinion, `AsyncKafkaConsumer` could provide better behavior, reducing the risk of `NPEs` for users, even if it introduces a small behavior change. We can document this change in the relevant methods.

Reply:
@chia7712—is your suggestion then to remove `null`s for both `beginningOffsets()` and `endOffsets()` and offsetsForTimes()` results?

Reply:
> is your suggestion then to remove nulls for both beginningOffsets() and endOffsets() and offsetsForTimes()` results?

Yes, it's not just my preference—Java 11 (and 4.0 clients) requires using Java 11, and most collection-related helpers don't accept null values.

I’ll open a JIRA to facilitate more discussion on this. For now, let's keep this PR focused on aligning the behavior.

Reply:
Because of the existing logic, it's not possible for the set of partitions passed to `beginningOffsets()` and `endOffsets()` to include partitions that are not in the returned set.

For example, if a topic has two partitions (0 and 1), attempting to search for partitions 0, 1, **and** 2 does not work. Wrong or right, both consumers will enter a loop, attempting to determine the leader for partition 2.

In that case, when the user calls `beginningOffsets()` or `endOffsets()` with a timeout, a `TimeoutException` will eventually be thrown. If no timeout is provided, those API calls hang forever looking for the leader 😱

Reply:
> I’ll open a JIRA to facilitate more discussion on this. For now, let's keep this PR focused on aligning the behavior.

Did you get a chance to create the new Jira?

Reply:
> If no timeout is provided, those API calls hang forever looking for the leader 

luckily, there is always a default timeout :smile: 

> Did you get a chance to create the new Jira?

Sorry, I completely forgot about that. I've opened: https://issues.apache.org/jira/browse/KAFKA-17826.